src/jenkins.py

```diff
@@ -705,3 +708,196 @@ def get_agent_name(unit_name: str) -> str:
+    plugin_dependency_outputs: typing.Iterable[str],
+) -> dict[str, tuple[str, ...]]:
+    """Build a lookup table of plugin short name to list of dependency plugin's short names.
+
+    Args:
+        plugin_dependency_outputs: The plugin dependency output from Jenkins Groovy script.
+
+    Returns:
+        The dependency lookup table.
+    """
+    dependency_lookup: dict[str, tuple[str, ...]] = {}
+    for line in plugin_dependency_outputs:
+        match = re.match(PLUGIN_LINE_CAPTURE, line)
+        if not match:
+            continue
+        plugin, dependencies = match.group(1), match.group(3)
+        if not dependencies:
+            dependency_lookup[plugin] = ()
+            continue
+        try:
+            dependency_lookup[plugin] = tuple(
+                _get_plugin_name(dependency) for dependency in dependencies.split(", ")
+            )
+        except ValidationError as exc:
+            logger.error("Invalid plugin dependency, %s", exc)
+            continue
+
+    return dependency_lookup
+
+
+def _set_jenkins_system_message(message: str, container: ops.Container) -> None:
+    """Set a system message on Jenkins.
+
+    Args:
+        message: The system message to display.
+        container: The Jenkins workload container.
+    """
+    jcasc_yaml = container.pull(JCASC_CONFIG_FILE_PATH, encoding="utf-8").read()
+    config = yaml.safe_load(jcasc_yaml)
+    config["jenkins"]["systemMessage"] = message
+    container.push(
+        JCASC_CONFIG_FILE_PATH, yaml.dump(config), encoding="utf-8", user=USER, group=GROUP
+    )
+
+
+def _traverse_dependencies(
+    plugin: str, dependency_lookup: typing.Mapping[str, typing.Iterable[str]], seen: set[str]
+) -> typing.Iterable[str]:
+    """Traverse through plugin dependencies that have not yet been traversed recursively.
+
+    Args:
+        plugin: The plugin to recurse through the dependencies.
+        dependency_lookup: The plugin and it's dependencies lookup table.
+        seen: a set indicating whether a plugin has already been traversed.
+
+    Yields:
+        Plugin and it's dependents.
+    """
+    if plugin in seen or plugin not in dependency_lookup:
+        return
+    yield plugin
+    seen.add(plugin)
+    for dependency in dependency_lookup.get(plugin, []):
+        yield from _traverse_dependencies(dependency, dependency_lookup, seen)
+
+
+def _get_allowed_plugins(
+    top_level_plugins: typing.Iterable[str],
+    dependency_lookup: typing.Mapping[str, typing.Iterable[str]],
+) -> typing.Iterable[str]:
+    """Get the plugin short names of allowed plugins and its dependencies.
+
+    Args:
+        top_level_plugins: The allowed plugins short names to add to allowed plugins with its
+            dependencies.
+        dependency_lookup: The plugin dependency lookup table.
+
+    Yields:
+        The allowed plugin short name.
+    """
+    seen: set[str] = set()
+    for plugin in top_level_plugins:
+        if plugin in seen:
```

Comment:
Why not
```
if not plugin in seen:
        yield from _traverse_dependencies(plugin, dependency_lookup, seen)
```
 ?

Reply:
I usually do the not continue case since I prefer the code not nested. I don't mind changing such simple cases though. Updating!

Reply:
Up to you. You don't have to change it